### PR TITLE
Toggle visibility of refs in the left panel

### LINKED ISSUE #5962: Option to show tags not stored on restart
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
On closing Git Extensions the option to show tags in the new left hand side branch overview does not get stored and get's reset to not checked

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

1. open a repository which includes tags
1. check `Show tags` found in the gear menu on the left hand panel
1. close Git Extensions and open it again
1. open any repository containing tags (might be the same as in the first step or another one)

(see gif)

gitexerror

**What is the expected behavior?**

The check box should stay checked after restart

**Environment you encounter the issue:**

- Git Extensions 3.00.00.4433
- Build fca7cf228b481ee8c1b779cf7b882ccdfbdcd1bc
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.3260.0
- DPI 96dpi (no scaling)

**Did this work in previous version of GitExtensions (which)?**
no, checkbox has been introduced in 3.x (as far as I know)

### PR DESCRIPTION
**Pull request by RussKie**
Part of #4814
Fixes #5962

## Proposed changes

* Replace cogs icon with a tree icon
* Allow toggle visibility of branches/remotes/tags
* Visibility state is persisted in user settings

### Before

image

### After

image

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
The current implementation will show a node as the last one in the tree, I'm somewhat struggling to come up with an elegant solution.
@amaiorano do you think you could help a little?

**Comment by amaiorano** [comment]
> The current implementation will show a node as the last one in the tree, I'm somewhat struggling to come up with an elegant solution.
> @amaiorano do you think you could help a little?

Hi @RussKie, I would be happy to help! But, would you be okay with putting this change on ice until I get my pub/sub rot change (https://github.com/gitextensions/gitextensions/pull/5669) done first? Just looking over your changes, I can see that it's going to be painful for me to rework it with my pubsub work, since the pubsub change consists mainly of replacing ReloadAsync at the RepoObjectsTree level with separate calls to a new ReloadNodes function that is invoked by each "top" node (i.e. Branches, Remotes, Tags, and eventually Submodules).

Already, I have to rework a change that went in recently (https://github.com/gitextensions/gitextensions/pull/5994) because the AheadBehind stuff should have been in RepoObjectsTree.Nodes.Branches.cs, not in RepoObjectsTree.cs. I'm trying to rework this as part of #5669, but I think I might have to make a separate PR for this, and rebase my work again on top of it.

So yeah, would you be cool with putting this one on the back burner for a bit? I actually started working again on #5669 just two days ago.

**Comment by RussKie** [comment]
> However the text labels get out of sync.

If a node is shown, the item is checked and the text should say "hide"; else - unticked and "show".
Is it not what you're seeing?

* [x] "Show branches" when branches are shown to me looked odd, and that prompted me to add alternate labels.

Would it help if items were not checked?

**Comment by gerhardol** [comment]
> If a node is shown, the item is checked and the text should say "hide"; else - unticked and "show".
Is it not what you're seeing?

Correct, 
I would prefer to not change the label though, just have the checkbox. "Show tags" is what is used in for instance View menu, I suggest that only 'Tags' is used in this situation.

**Comment by RussKie** [comment]
Ok, simplify items labels to "Branches", "Tags", etc.
Sounds sensible, will update

**Comment by RussKie** [comment]
@gerhardol  I've updated the labels.

@amaiorano please let me know when it is a good time to resurrect this. 
...or feel free to take over this piece (and close it).

**Comment by amaiorano** [comment]
> @amaiorano please let me know when it is a good time to resurrect this.
> ...or feel free to take over this piece (and close it).

Since my #5669 change has been merged, you can go ahead and resurrect this. I'm not sure I'll be able to take it over, though. Am presently swamped, and trying to get my left panel submodules change in, which is a series of me breaking it down into smaller PRs.

**Comment by amaiorano** [comment]
Am taking a crack at this, @RussKie.

**Comment by amaiorano** [comment]
@RussKie take a look at the rebased commits. I patched up the first commit so that it makes use of the new RefreshTree code I added.

The 3rd commit makes sure to disable the entire left panel while reloading nodes, which fixes issues with toggling visibility of top nodes while refreshing, which would lead to NullReferenceExceptions. Disabling the entire panel is much safer.

The 4th commit is my crack at fixing the ordering problem. It's pretty straightforward, and does the trick. I think this should also allow us to make the order configurable easily enough.

Anyway, take it for a spin.

**Comment by RussKie** [comment]
My apologies, I have started the rebase and tweaks but I have run out of time to push them in.
Thank you for finishing it up

File: GitUI/BranchTreePanel/RepoObjectsTree.cs [hunk 1]
```diff
@@ -201,53 +211,89 @@ protected override void OnUICommandsSourceSet(IGitUICommandsSource source)
...
+                SelectedImageKey = nameof(Images.TagHorizontal),
+            };
             _tagTree = new TagTree(_tagTreeRootNode, UICommandsSource);
-            AddTree(_tagTree);
+            AddTree(_tagTree, 2);
             _searchResult = null;
         }
 
-        private void AddTree(Tree tree)
+        private Dictionary<Tree, int> _treeToPositionIndex = new Dictionary<Tree, int>();
```
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs]: this need to move to the top with the rest of fields

**Comment by RussKie** [comment]
It works. 👍 

However there is an annoying flicker when a node is added back to the tree.

**Comment by RussKie** [comment]
One more annoying behaviour - when the branches node is added it is automatically expanded, even though it may have been collapsed before hiding

**Comment by amaiorano** [comment]
> One more annoying behaviour - when the branches node is added it is automatically expanded, even though it may have been collapsed before hiding

I think this is totally fine. If a user hides a node, they are opting into not seeing it anymore, likely for a long time. Adding it back and having if reset to it's default state (expanded for Branches, collapses for Tags) is fine. It's only annoying to you because you're testing it's behavior.

**Comment by amaiorano** [comment]
Btw, good catch on the double refresh!

**Comment by amaiorano** [comment]
> My apologies, I have started the rebase and tweaks but I have run out of time to push them in.
> Thank you for finishing it up

No worries! It's fun working together :)

**Pull request merged**